File: .claude/skills/flow-engineer-command/scripts/package_command.ts
#!/usr/bin/env -S deno run -A
/**
 * Command Packager - Creates a distributable .skill file of a command folder
 *
 * Usage:
 *     deno run -A package_command.ts <path/to/command-folder> [output-directory]
 *
 * Example:
 *     deno run -A package_command.ts skills/public/flow-my-command
 *     deno run -A package_command.ts skills/public/flow-my-command ./dist
 */

import { basename, join, relative, resolve } from "jsr:@std/path";
import { validateCommand } from "./validate_command.ts";
import { walkSync } from "jsr:@std/fs/walk";

function packageCommand(
  commandPathArg: string,
  outputDir: string | null = null,
): string | null {
  const commandPath = resolve(commandPathArg);

  // Validate command folder exists
  try {
    const stat = Deno.statSync(commandPath);
    if (!stat.isDirectory) {
      console.log(`❌ Error: Path is not a directory: ${commandPath}`);
      return null;
    }
  } catch {
    console.log(`❌ Error: Command folder not found: ${commandPath}`);
    return null;
  }

  // Validate SKILL.md exists
  const skillMd = join(commandPath, "SKILL.md");
  try {
    Deno.statSync(skillMd);
  } catch {
    console.log(`❌ Error: SKILL.md not found in ${commandPath}`);
    return null;
  }

  // Run validation before packaging
  console.log("🔍 Validating command...");
  const [valid, message] = validateCommand(commandPath);
  if (!valid) {
    console.log(`❌ Validation failed: ${message}`);
    console.log("   Please fix the validation errors before packaging.");
    return null;
  }
  console.log(`✅ ${message}\n`);

  // Determine output location
  const commandName = basename(commandPath);
  let outputPath: string;
  if (outputDir) {
    outputPath = resolve(outputDir);
    Deno.mkdirSync(outputPath, { recursive: true });
  } else {
    outputPath = Deno.cwd();
  }

  const commandFilename = join(outputPath, `${commandName}.skill`);

  // Create the .skill file (zip format) using the `zip` command
  try {
    // Collect all files relative to parent directory
    const parentDir = resolve(commandPath, "..");
    const entries: string[] = [];

    for (const entry of walkSync(commandPath, { includeDirs: false })) {
      const arcname = relative(parentDir, entry.path);
      entries.push(arcname);
    }

    // Sort entries for deterministic output (Python's rglob returns sorted on most platforms)
    entries.sort();

    // Use zip command to create the archive
    // First, remove existing file if present
    try {
      Deno.removeSync(commandFilename);
    } catch {
      // File may not exist, that's fine
    }

    // Use zip -r with the command directory name from the parent dir
    // to preserve directory structure inside the archive.
    const zipCmd = new Deno.Command("zip", {
      args: ["-r", commandFilename, commandName],
      cwd: parentDir,
      stdout: "null",
      stderr: "piped",
    });

    const zipResult = zipCmd.outputSync();
    if (!zipResult.success) {
      const stderr = new TextDecoder().decode(zipResult.stderr);
      console.log(
        `❌ Error creating .skill file: zip command failed: ${stderr}`,
      );
      return null;
    }

    // Print added files (matching Python output format: relative to parent)
    for (const arcname of entries) {
      console.log(`  Added: ${arcname}`);
    }

    console.log(`\n✅ Successfully packaged command to: ${commandFilename}`);
    return commandFilename;
  } catch (e) {
    console.log(`❌ Error creating .skill file: ${e}`);
    return null;
  }
}

function main(): void {
  if (Deno.args.length < 1) {
    console.log(
      "Usage: deno run -A package_command.ts <path/to/command-folder> [output-directory]",
    );
    console.log("\nExample:");
    console.log(
      "  deno run -A package_command.ts skills/public/flow-my-command",
    );
    console.log(
      "  deno run -A package_command.ts skills/public/flow-my-command ./dist",
    );
    Deno.exit(1);
  }

  const commandPath = Deno.args[0];
  const outputDir = Deno.args.length > 1 ? Deno.args[1] : null;

  console.log(`📦 Packaging command: ${commandPath}`);
  if (outputDir) {
    console.log(`   Output directory: ${outputDir}`);
  }
  console.log();

  const result = packageCommand(commandPath, outputDir);

  if (result) {
    Deno.exit(0);
  } else {
    Deno.exit(1);
  }
}

if (import.meta.main) {
  main();
}

File: .claude/skills/flow-engineer-command/scripts/package_command.py
#!/usr/bin/env python3
"""
Command Packager - Creates a distributable .skill file of a command folder

Usage:
    python utils/package_command.py <path/to/command-folder> [output-directory]

Example:
    python utils/package_command.py skills/public/flow-my-command
    python utils/package_command.py skills/public/flow-my-command ./dist
"""

import sys
import zipfile
from pathlib import Path
from validate_command import validate_command


def package_command(command_path, output_dir=None):
    """
    Package a command folder into a .skill file.

    Args:
        command_path: Path to the command folder
        output_dir: Optional output directory for the .skill file (defaults to current directory)

    Returns:
        Path to the created .skill file, or None if error
    """
    command_path = Path(command_path).resolve()

    # Validate command folder exists
    if not command_path.exists():
        print(f"❌ Error: Command folder not found: {command_path}")
        return None

    if not command_path.is_dir():
        print(f"❌ Error: Path is not a directory: {command_path}")
        return None

    # Validate SKILL.md exists
    skill_md = command_path / "SKILL.md"
    if not skill_md.exists():
        print(f"❌ Error: SKILL.md not found in {command_path}")
        return None

    # Run validation before packaging
    print("🔍 Validating command...")
    valid, message = validate_command(command_path)
    if not valid:
        print(f"❌ Validation failed: {message}")
        print("   Please fix the validation errors before packaging.")
        return None
    print(f"✅ {message}\n")

    # Determine output location
    command_name = command_path.name
    if output_dir:
        output_path = Path(output_dir).resolve()
        output_path.mkdir(parents=True, exist_ok=True)
    else:
        output_path = Path.cwd()

    command_filename = output_path / f"{command_name}.skill"

    # Create the .skill file (zip format)
    try:
        with zipfile.ZipFile(command_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:
            # Walk through the command directory
            for file_path in command_path.rglob('*'):
                if file_path.is_file():
                    # Calculate the relative path within the zip
                    arcname = file_path.relative_to(command_path.parent)
                    zipf.write(file_path, arcname)
                    print(f"  Added: {arcname}")

        print(f"\n✅ Successfully packaged command to: {command_filename}")
        return command_filename

    except Exception as e:
        print(f"❌ Error creating .skill file: {e}")
        return None


def main():
    if len(sys.argv) < 2:
        print("Usage: python utils/package_command.py <path/to/command-folder> [output-directory]")
        print("\nExample:")
        print("  python utils/package_command.py skills/public/flow-my-command")
        print("  python utils/package_command.py skills/public/flow-my-command ./dist")
        sys.exit(1)

    command_path = sys.argv[1]
    output_dir = sys.argv[2] if len(sys.argv) > 2 else None

    print(f"📦 Packaging command: {command_path}")
    if output_dir:
        print(f"   Output directory: {output_dir}")
    print()

    result = package_command(command_path, output_dir)

    if result:
        sys.exit(0)
    else:
        sys.exit(1)


if __name__ == "__main__":
    main()
